op-batcher/batcher/espresso.go

```diff
@@ -39,140 +35,285 @@ const (
-		l.recordFailedDARequest(txdata.ID(), err)
-		return nil, fmt.Errorf("failed to submit transaction: %w", err)
+func (l *BatchSubmitter) espressoSyncAndRefresh(ctx context.Context, newSyncStatus *eth.SyncStatus, streamer *espresso.EspressoStreamer) {
+	shouldClearState, err := streamer.Refresh(ctx, newSyncStatus)
+	shouldClearState = shouldClearState || err != nil
+
+	l.channelMgrMutex.Lock()
+	defer l.channelMgrMutex.Unlock()
+	syncActions, outOfSync := computeSyncActions(*newSyncStatus, l.prevCurrentL1, l.channelMgr.blocks, l.channelMgr.channelQueue, l.Log, l.Config.PreferLocalSafeL2)
+	if outOfSync {
+		l.Log.Warn("Sequencer is out of sync, retrying next tick.")
+		return
 	}
+	l.prevCurrentL1 = newSyncStatus.CurrentL1
+	if syncActions.clearState != nil || shouldClearState {
+		l.channelMgr.Clear(*syncActions.clearState)
+	} else {
+		l.channelMgr.PruneSafeBlocks(syncActions.blocksToPrune)
+		l.channelMgr.PruneChannels(syncActions.channelsToPrune)
+	}
+}
 
-	timer := time.NewTimer(transactionFetchTimeout)
-	defer timer.Stop()
+// Periodically refreshes the sync status and polls Espresso streamer for new batches
+func (l *BatchSubmitter) espressoBatchLoadingLoop(ctx context.Context, wg *sync.WaitGroup, publishSignal chan struct{}) {
+	l.Log.Info("Starting EspressoBatchLoadingLoop")
 
-	ticker := time.NewTicker(transactionFetchInterval)
+	defer wg.Done()
+	ticker := time.NewTicker(l.Config.PollInterval)
 	defer ticker.Stop()
 
-	var txQueryData espressoCommon.TransactionQueryData
-Loop:
+	streamer := espresso.EspressoStreamer{
+		BatcherAddress: l.BatcherAddress,
+		Namespace:      l.RollupConfig.L2ChainID.Uint64(),
+
+		L1Client:            l.L1Client,
+		EspressoClient:      l.Espresso,
+		EspressoLightClient: l.EspressoLightClient,
+		Log:                 l.Log,
+
+		BatchPos: 1,
+	}
+
 	for {
 		select {
 		case <-ticker.C:
-			txQueryData, err = l.Espresso.FetchTransactionByHash(l.shutdownCtx, txHash)
-			if err == nil {
-				break Loop
+			newSyncStatus, err := l.getSyncStatus(ctx)
+			if err != nil {
+				l.Log.Error("failed to refresh sync status", "err", err)
+				continue
 			}
-			l.Log.Warn("Retry fetching transaction by hash", "txHash", txHash, "error", err)
-		case <-timer.C:
-			l.Log.Error("Failed to fetch transaction by hash after multiple attempts", "txHash", txHash)
-			l.recordFailedDARequest(txdata.ID(), err)
-			return nil, fmt.Errorf("failed to fetch transaction by hash: %w", err)
+
+			l.espressoSyncAndRefresh(ctx, newSyncStatus, &streamer)
+
+			err = streamer.Update(ctx)
+			if err != nil {
+				l.Log.Error("failed to update Espresso streamer", "err", err)
+				continue
+			}
+
+			var batch *espresso.EspressoBatch
+			for {
+				batch = streamer.Next(ctx)
+				if batch == nil {
+					break
+				}
+
+				block, err := espresso.BatchToIncompleteBlock(l.RollupConfig, batch)
+				if err != nil {
+					l.Log.Error("failed to convert singular batch to block", "err", err)
+					continue
```

Comment:
Continue here causes the batch to be skipped. So we need to be sure we only do this if the failure is deterministic (ie if the batch is fundamentally not convertible to a block, due to being malformed or something). If `BatchToIncompleteBlock` could fail for some transient reason (idk, failure to talk to L1 or something like that) this could be a problem.

I don't know how `BatchToIncompleteBlock` works so maybe this is fine, but it's at least worth a comment flagging this as a high risk section, especially if `BatchToIncompleteBlock` could change in the future

Reply:
Makes sense, I've added corresponding comments to both the function declaration and this section.
The function will indeed fail only if the batch is malformed or if the batcher is misconfigured, in which case it will fail always, which I don't think should be a problem here.